apps/os/backend/agent/keep-alive-agent.ts

```diff
@@ -0,0 +1,340 @@
+              `${options.logErrorAfter ? `If this is expected, provide a logErrorAfter value to betterWaitUntil to indicate when to log an error on unresovled promises.` : ""}` +
+              `${!options.timeout ? ` You can provide a timeout value to betterWaitUntil that will stop promises from being waited on forever.` : `This promise will terminate at approximately ${options.timeout.toISOString()}`}`,
+          );
+          lastLoggedErrorAt = Date.now();
+        }
+
+        if (options.timeout && Date.now() > options.timeout.getTime()) {
+          // do not use debug for this because we want it to surface
+          logger.error(
+            "[better-wait-until] Timeout reached, stopping keep alive interval. Your Durable Object may now be killed by Cloudflare and the promise may never resolve.",
+          );
+          clearInterval(interval);
+          resolve();
+          return;
+        }
+
+        // Cloudflare sometimes gets funky with Date.now outside of a request context so we record the iteration count as well
+        logDebug(
+          `Background task has been running for ${Date.now() - start}ms (iteration ${count})`,
+        );
+
+        const response = await websocketPromise;
+        response.webSocket!.send("better-wait-until-ping " + count);
+      } catch (err) {
+        logError("Error keeping agent awake", { err });
+      }
+    }, 10000);
+  });
+
+  return intervalFinished;
+}
+
+export function betterWaitUntil(
+  durableObject: DurableObject,
+  promise: Promise<unknown>,
+  options: {
+    fetchOptions?: Parameters<typeof fetch>[1];
+    timeout?: Date;
+    logWarningAfter?: Date;
+    logErrorAfter?: Date;
+  } = {},
+): void {
+  void betterAwait(durableObject, promise, options);
+}
+
+// export under `waitUntil` for autocomplete
+export { betterWaitUntil as waitUntil };
+// export as default for convenience
+export default betterWaitUntil;
+
+// Agents-specific code
+
+function agentsConstructorUpdate(instance: Agent<any, any>): void {
+  const originalBroadcast = instance.broadcast;
+  Object.defineProperty(instance, "broadcast", {
+    configurable: true,
+    enumerable: false,
+    writable: true,
+    value: function (this: Agent<any, any>, msg: string, without?: string[]) {
+      const exclude: string[] = without ?? [];
+      for (const connection of this.getConnections()) {
+        if (connection.url?.includes("better-wait-until/websocket")) {
+          logDebug("Skipping keep-alive connection", connection);
+          exclude.push(connection.id);
+        }
+      }
+      return originalBroadcast.call(this, msg, exclude);
+    },
+  });
+  // this is supersitious, I don't think it's needed
+  try {
+    instance["_autoWrapCustomMethods"]();
+  } catch (error) {
+    logError("Error wrapping custom methods", { error });
+  }
+}
+
+export abstract class KeepAliveAgent<
```

Comment:
I still think that an abstract class that magically modifies properties (and controls which `Agent` base class we use) is less good than a function that we have to call from the constructor that looks like `monkeyPatchAgentPrototype(this)`. I guess basically like you have it here, but we would move the stuff in this class directly into `IterateAgent` so it's explicit. Without that, I think `KeepAliveAgent` is going to a) be unfairly accused of causing all our problems or b) be fairly accused of causing all our problems (less likely)

Reply:
will do